mpf/config_players/blinkenlight_player.py

```diff
@@ -0,0 +1,67 @@
+"""Blinkenlight config player."""
+import uuid
+
+from mpf.config_players.device_config_player import DeviceConfigPlayer
+
+
+class BlinkenlightPlayer(DeviceConfigPlayer):
+
+    """Blinks lights different colors based on config."""
+
+    config_file_section = 'blinkenlight_player'
+    show_section = 'blinkenlights'
+    machine_collection_name = 'blinkenlights'
+    allow_placeholders_in_keys = True
+
+    def play(self, settings, context, calling_context, priority=0, **kwargs):
+        """Adds/removes colors to blink."""
+        del kwargs
+
+        for blinkenlight, s in settings.items():
```

Comment:
This is kind of different than most of our other players. Not sure if better or worse. One issue might be consistency.

Usually you can add and remove stuff with or without a `key`. Here this is called `label`. However, here you kind of have to use a label or removing your entry will not work later on (because you do not know the uuid). Also what you can do here is to remove `labels` from other modes. `keys` are namespaced in other players so that you cannot have conflicts between modes. Then we got contexts and labels here. In other places we just map remember our keys insides a context dict. Here `blinkenlight` needs to know both `context` and `label`.

Any thoughts on this? I'm always open to suggestions for improvement.

Reply:
So is your main concern that a mode can remove a color that was added by another mode if it uses the same label? My thought there is, that's kind of intended. Honestly, I think most of the time the user won't even use a `label` since the color is automatically removed when the mode ends. But it could be that you want to remove the color when a shot is hit or something.

Reply:
> So is your main concern that a mode can remove a color that was added by another mode if it uses the same label? My thought there is, that's kind of intended. Honestly, I think most of the time the user won't even use a `label` since the color is automatically removed when the mode ends. But it could be that you want to remove the color when a shot is hit or something.

I see that this might be a feature but it might as well confuse people. Conceptually, I think that it would be bad to design something which is added in one mode and removed in another one. That will lead to confusion and issues down the road.

In my opinion the main usecase would be this (which currently will not work this way):

```
blinkenlight_player:
  mode_my_mode_started:
    my_blinkenlight1: orange
  some_other_event:
    my_blinkenlight1: remove/stop    # some magic color. we use "stop" for lights
  some_other_event2:
    my_blinkenlight1:
      action: remove
```

Reply:
As this is being discussed here, what is the future possibility of integrating this as a show?  I am thinking a shot profile call a show based on state.  Can I have a show for state 1, which adds to a blinkenlight with priority?  Then when it goes to an "off" state, I remove that key from the blinkenlight?

I don't think that would work today, and I don't think its necessary.  But for broader appeal and use, that would be a nice way to be able to call and use this feature.

Reply:
That should already work. Your profile will stop and start shows. Stop will call clear_context and remove the previous color. Guess we should add a test with a show but I am quite confident that it will just work.

Reply:
Just to make this clear: I am very happy with the code how it is. I would like to change `label` to `key` for consistency and unexpose the context. Thats about it. Config interface is hard to change later. Refactoring the internal code can be done later and nobody will notice.

Reply:
Awesome. Well I can play around with it as I'm thinking and add tests as we go.
And as Jan said, awesome job! I think this will be a popular feature!

Reply:
Hmm. Something to consider if we go with what you're suggesting. What happens if a mode adds several colors (not sure why a mode would but it's totally possible and is designed to allow this). Then what would happen if:

```
blinkenlight_player:
  some_event:
    my_bl: remove
```

Would this remove all colors from the mode? Just the latest one?

Reply:
> As this is being discussed here, what is the future possibility of integrating this as a show?  I am thinking a shot profile call a show based on state.  Can I have a show for state 1, which adds to a blinkenlight with priority?  Then when it goes to an "off" state, I remove that key from the blinkenlight?
> 
> 
> 
> I don't think that would work today, and I don't think its necessary.  But for broader appeal and use, that would be a nice way to be able to call and use this feature.

This wouldn't quite work because the blinkenlight will remove the show's colors when the show ends. So if your shot profile called a show which added a color, that color would only exist for the duration of the show.

Reply:
Yeah, that's what I wasn't sure about. If I had a show similar to on, then it would never end. So theoretically it would be in that state until I told it to end that show, which would remove it.

Otherwise that was my though, I add it when mode starts, and then xx_lit_hit would be my trigger to remove it. I'm excited to test this through the scenarios!

Reply:
> I add it when mode starts, and then xx_lit_hit would be my trigger to remove it. I'm excited to test this through the scenarios!

Yep. Just keep in mind that (as of now) if you want to remove the color yourself (as opposed to automatically removing the color when the mode ends) then you'll need to specify a `label` value when you add the color, then `remove` using that label later.

Reply:
Yep, and that's consistent with keys to remove widgets and other players. So people should be able to follow that logic through to here with a nice write up in the docs. I'm more than willing to assist with that piece for nice we get the code finalized.

Reply:
You can check out my PR on the mpf-docs repo. I've already added documentation for blinkenlight in that pull request. I'm getting a Travis CI error though and I'm not sure why.

Reply:
> Hmm. Something to consider if we go with what you're suggesting. What happens if a mode adds several colors (not sure why a mode would but it's totally possible and is designed to allow this). Then what would happen if:
> 
> ```
> blinkenlight_player:
>   some_event:
>     my_bl: remove
> ```
> 
> Would this remove all colors from the mode? Just the latest one?

In this case you should use keys. Otherwise the second color should remove the first one (because they got the same key). That is the logic for all other players at least. Except maybe in widget player where you can intentionally leak objects to add one widget multiple times (prone to memory leaks obviously).

> > As this is being discussed here, what is the future possibility of integrating this as a show?  I am thinking a shot profile call a show based on state.  Can I have a show for state 1, which adds to a blinkenlight with priority?  Then when it goes to an "off" state, I remove that key from the blinkenlight?
> > I don't think that would work today, and I don't think its necessary.  But for broader appeal and use, that would be a nice way to be able to call and use this feature.
> 
> This wouldn't quite work because the blinkenlight will remove the show's colors when the show ends. So if your shot profile called a show which added a color, that color would only exist for the duration of the show.

Basically, your color has to be included in all shows where it should exist. Similar to light_player behaviour.

Reply:
Played around with this and so far with basic tests its working great.  I tried adding it to shot_profiles, and it worked just as Jan said it would.  When I dug in, I did run into one error though.  Placeholders for blinkenlights work fine, but for color threw an error.
```
shows:
    bl_on:
      - duration: -1
        blinkenlights:
            (light): (color_1)

File "c:\users\atummons\desktop\davebranch\mpf\mpf\core\config_validator.py", line 683, in _validate_type_color return int(color[0]), int(color[1]), int(color[2])
ValueError: invalid literal for int() with base 10: '(color_1)'
```
Not sure why this works for shows with lights, but not blinkenlights.  I looked around, but don't really know what I was looking for.

Reply:
Good find. I think we need color_or_token as type to support that.

Reply:
I made the comment in the config spec.  Didn't need to make it a token.  But in config spec it was a type color, but needs to be a type str.  This is the same as light_player.  I made the change local, and it worked as expected.  So I made a comment on that row for the change.

Not sure what's appropriate and good coding practices.  Should I make a push for that change to someone else's branch?  Or just note it for them?  Not sure what standard practice is, or if its all personal preference for Dave, since this is his branch.

Reply:
Go ahead and make the change if you want. I was holding off on making changes in case Jan decided to do a stream today, but generally I have no problems if you want to make changes.

Reply:
Maybe we can finish this baby together later on. Should be fun.

Reply:
Its looking super awesome.  I put it through a few more paces this morning, and worked just as expected!  So I think a few of those cleanup things from this thread, and if we want to implement the color_or_token method and we should be in business!

Reply:
Woohoo!

Reply:
I put together a fix for a method call that was wrong, and added color_or_token, but saying I'm getting permission denied to push to Dave's branch.  So is best option to just post changes here and someone with authority can make them?  Or is there a way I can push against it and I'm trying to push to Dave's branch, instead of this PR?

Reply:
> I put together a fix for a method call that was wrong, and added color_or_token, but saying I'm getting permission denied to push to Dave's branch.  So is best option to just post changes here and someone with authority can make them?  Or is there a way I can push against it and I'm trying to push to Dave's branch, instead of this PR?

I have it set to allow changes from mpf maintainers (if you have write access to mpf you can change the PR) but I don't think I can allow changes from anyone else? My GitHub knowledge is lacking here lol

Reply:
Yeah, I'm not an mpf maintainer, I don't believe.  I have your branch checked out on local, and am pushing to your local.  I think you have to add me to your repo as a collaborator (don't do this, I have limited knowledge as well and am not to be trusted on your branches!).  So I can shoot it over (its only about 10 lines) if we get together to discuss final stuff.

Reply:
> I would like to change `label` to `key` for consistency and unexpose the context.

I renamed `label` back to `key`.  As far as "unexposed the context" I'm not sure what you would like to do there, exactly.  So I'll let you make those changes, if that's ok.

Reply:
Pulled this in and all is working as expected with both of your updates.